### PR TITLE
[wdspec] rewrite remove network interception tests to fail gracefully

### PR DESCRIPTION
**Pull request by thiagowfx**
- ~~Use navigation with an expected timeout instead of fetch with an expected exception.~~

~~From a test implementation perspective this is a no-op as, when the network interception triggers, the navigation is expected to not complete (and thus timeout).~~

(moved this part to its own PR: https://github.com/web-platform-tests/wpt/pull/42969)

- Introduce a "wait_for_future_safe" fixture that gracefully fails the await of a future instead of making the whole test suite fail.

Instead of aborting the "wpt run" when there is a timeout, treat the timeout as a failure, effectively failing gracefully. This allows subsequent tests to run, and expectations to be updated later on.

Related: https://github.com/w3c/webdriver-bidi/issues/188

Bug: #42482

### CODE REVIEW

**Comment by thiagowfx** [comment]
Worth to discuss:

- bikeshed on the name of the helper fixture: `wait_for_future_safe`, `wait_for_future_with_timeout`, etc
- maybe we could add a custom exception / error message to the fixture to make the failure reason more obvious

**Comment by thiagowfx** [comment]
I tested this change locally with the mapper and it works well!

There are 3 tests: "test_remove_intercept[beforeRequestSent]", "test_remove_intercept[responseStarted]", "test_return_value".

"test_remove_intercept[responseStarted]" is failing for Chrome. This is our problem, we'll fix it later. That said:

## Currently

The failure blocks the whole "wpt run" execution. No expectations are updated. Then we are forced to fully disable all tests in the remove_intercept directory: https://github.com/GoogleChromeLabs/chromium-bidi/blob/main/wpt-metadata/mapper/headless/webdriver/tests/bidi/network/remove_intercept/remove_intercept.py.ini

It's all or nothing, because we cannot expect subtests to handle timeout expectations properly with the current runner infra / implementation (see https://github.com/web-platform-tests/wpt/issues/42482).

## With this PR

The first test passes, then the second test FAILS instead of TIMING OUT, then the third test passes, then the WPT expectations are properly updated for the second test.

File: webdriver/tests/support/fixtures_bidi.py [hunk 1]
```diff
@@ -101,21 +102,35 @@ async def on_event(_, data):
...
+
 @pytest.fixture
 def current_time(bidi_session, top_context):
     """Get the current time stamp in ms from the remote end.
 
     This is required especially when tests are run on different devices like
     for Android, where it's not guaranteed that both machines are in sync.
     """
-    async def _():
+    async def current_time():
```
Comment (thiagowfx) [inline webdriver/tests/support/fixtures_bidi.py:126]: drive-by refactoring, for consistency

**Comment by thiagowfx** [comment]
cc @Lightning00Blade

**Review by OrKoN — APPROVED** [review]

**Comment by thiagowfx** [comment]
cc @juliandescottes @whimboo

**Comment by thiagowfx** [comment]
I'd really like to move forward with the pattern in this PR, it will help unblock us for pretty much everything related to the network module that awaits events. Do you have some initial feedback, or can I go ahead and merge it already?

File: webdriver/tests/bidi/network/remove_intercept/remove_intercept.py [hunk 2]
```diff
@@ -39,12 +41,15 @@ async def test_remove_intercept(
...
-    # from the AbortController.
-    with pytest.raises(ScriptEvaluateResultException):
-        await fetch(text_url)
+    # Request to top_context should be blocked.
+    # TODO(https://github.com/w3c/webdriver-bidi/issues/188): Use a timeout argument when available.
+    with pytest.raises(asyncio.TimeoutError):
+        await asyncio.wait_for(
+            asyncio.shield(bidi_session.browsing_context.navigate(context=top_context["context"], url=text_url, wait="complete")),
+            timeout=2.0,
+        )
```
Comment (sadym-chromium) [inline webdriver/tests/bidi/network/remove_intercept/remove_intercept.py:50]: doesn't this add 2 sec for the test runtime?

Comment (thiagowfx) [reply webdriver/tests/bidi/network/remove_intercept/remove_intercept.py:50]: Yes.

Comment (sadym-chromium) [reply webdriver/tests/bidi/network/remove_intercept/remove_intercept.py:50]: Instead, we can subscribe to some browsing-context specific event (?`browsingContext.load`?), and verify it was not emitted.

Comment (thiagowfx) [reply webdriver/tests/bidi/network/remove_intercept/remove_intercept.py:50]: Note that this PR is doing two different things: one is to introduce (and use) `wait_for_future_safe`, the other one is to rewrite the remove intercept test to depend on timeouts instead of exceptions (we discussed this in Matrix).

In case you have a concern with the latter but not with the former, I could split the latter into its own PR.

Comment (thiagowfx) [reply webdriver/tests/bidi/network/remove_intercept/remove_intercept.py:50]: > and verify it was not emitted.

Wouldn't this also depend on a timeout?

File: webdriver/tests/support/fixtures_bidi.py [hunk 3]
```diff
@@ -101,21 +103,38 @@ async def on_event(_, data):
         remove_listener()
 
 
+@pytest.fixture
+def wait_for_future_safe(configuration):
```
Comment (sadym-chromium) [inline webdriver/tests/support/fixtures_bidi.py:107]: This part LGTM

**Review by sadym-chromium — CHANGES_REQUESTED** [review]
`wait_for_future_safe` and it's use LGTM.
Waiting for 2 sec in each test is arguable.

File: webdriver/tests/support/fixtures_bidi.py (hunk 3 above)
Comment (thiagowfx) [reply webdriver/tests/support/fixtures_bidi.py:107]: Ack

File: webdriver/tests/bidi/network/remove_intercept/remove_intercept.py (hunk 2 above)
Comment (thiagowfx) [reply webdriver/tests/bidi/network/remove_intercept/remove_intercept.py]: Reverted. Splitting into another PR.

**Review by sadym-chromium — APPROVED** [review]

File: webdriver/tests/support/fixtures_bidi.py (hunk 3 above)
Comment (whimboo) [inline webdriver/tests/support/fixtures_bidi.py:107]: I assume that we can also use this new fixture for the two usages of `asyncio.shield` under `browsing_context` as well? If yes, feel free to replace so that we don't forget about it.

**Review by whimboo — APPROVED** [review]
Thank you!

File: webdriver/tests/support/fixtures_bidi.py (hunk 3 above)
Comment (thiagowfx) [reply webdriver/tests/support/fixtures_bidi.py:107]: I think so too. I am planning to send another PR that uses this fixture in all places where it's possible, starting with the network module.

**Comment by thiagowfx** [comment]
@DanielRyanSmith could you please merge? Thanks!
